File: CHAPTER 7. 동작에서 계산으로/yckim/chapter7/java/v1/InMemoryTrips.java
package chapter7.java.v1;

import java.time.Clock;
import java.util.*;
import java.util.stream.Collectors;

public class InMemoryTrips implements Trips {

    private final Clock clock;

    private final Map<String, Set<Trip>> trips = new HashMap<>();

    public InMemoryTrips(Clock clock) {
        this.clock = clock;
    }

    public void addTrip(Trip trip) {
        var existingTrips = trips.getOrDefault(
                trip.getCustomerId(),
                new HashSet<>()
        );
        existingTrips.add(trip);
        trips.put(trip.getCustomerId(), existingTrips);
    }

    @Override
    public Set<Trip> tripsFor(String customerId) {
        return trips.getOrDefault(customerId, Collections.emptySet());
    }

    @Override
    public Set<Trip> currentTripsFor(String customerId) {
        var now = clock.instant();
        return tripsFor(customerId).stream()
                .filter(trip -> trip.isPlannedToBeActiveAt(now))
                .collect(Collectors.toSet());
    }

}

File: CHAPTER 7. 동작에서 계산으로/yckim/chapter7/kotlin/v1/InMemoryTrips.kt
package chapter7.kotlin.v1

import chapter7.java.v2.Trip
import chapter7.java.v2.Trips
import java.time.Clock
import java.time.Instant
import java.util.stream.Collectors

class InMemoryTrips(private val clock: Clock) : Trips {
    private val trips: MutableMap<String, MutableSet<Trip>> = HashMap()
    fun addTrip(trip: Trip) {
        val existingTrips = trips.getOrDefault(
            trip.customerId,
            HashSet()
        )
        existingTrips.add(trip)
        trips[trip.customerId] = existingTrips
    }

    override fun tripsFor(customerId: String): Set<Trip> {
        return trips.getOrDefault(customerId, emptySet())
    }

    override fun currentTripsFor(customerId: String, at: Instant): Set<Trip> {
        return tripsFor(customerId).stream()
            .filter { trip: Trip -> trip.isPlannedToBeActiveAt(at) }
            .collect(Collectors.toSet())
    }
}